lib/logstash/outputs/kusto/ingestor.rb

```diff
@@ -78,41 +78,58 @@ def upload_async(path, delete_on_success)
-      @logger.error('Uploading failed, retrying.', exception: e.class, message: e.message, path: path, backtrace: e.backtrace)
-      sleep RETRY_DELAY_SECONDS
-      retry
+      begin
+        file_size = File.size(path)
+        @logger.debug("Sending file to kusto: #{path}. size: #{file_size}")
+
+        # TODO: dynamic routing
+        # file_metadata = path.partition('.kusto.').last
+        # file_metadata_parts = file_metadata.split('.')
+
+        # if file_metadata_parts.length == 3
+        #   # this is the number we expect - database, table, json_mapping
+        #   database = file_metadata_parts[0]
+        #   table = file_metadata_parts[1]
+        #   json_mapping = file_metadata_parts[2]
+
+        #   local_ingestion_properties = Java::KustoIngestionProperties.new(database, table)
+        #   local_ingestion_properties.addJsonMappingName(json_mapping)
+        # end
+
+        file_source_info = Java::com.microsoft.azure.kusto.ingest.source.FileSourceInfo.new(path, 0); # 0 - let the sdk figure out the size of the file
+        @kusto_client.ingestFromFile(file_source_info, @ingestion_properties)
+
+        @logger.debug("File #{path} sent to kusto.")
+      rescue Errno::ENOENT => e
+        @logger.error("File doesn't exist! Unrecoverable error.", exception: e.class, message: e.message, path: path, backtrace: e.backtrace)
+      rescue Java::JavaNioFile::NoSuchFileException => e
+        @logger.error("File doesn't exist! Unrecoverable error.", exception: e.class, message: e.message, path: path, backtrace: e.backtrace)
+      rescue => e
+        # When the retry limit is reached or another error happen we will wait and retry.
+        #
+        # Thread might be stuck here, but I think its better than losing anything
+        # its either a transient errors or something bad really happened.
+        @logger.error('Uploading failed, retrying.', exception: e.class, message: e.message, path: path, backtrace: e.backtrace)
+        sleep RETRY_DELAY_SECONDS
+        retry
+      end
+
+      begin
+        File.delete(path) if delete_on_success
+        @logger.debug("File #{path} deleted.")
+      rescue Errno::ENOENT => e
+        @logger.error("File doesn't exist! Unrecoverable error.", exception: e.class, message: e.message, path: path, backtrace: e.backtrace)
+      rescue Java::JavaNioFile::NoSuchFileException => e
+        @logger.error("File doesn't exist! Unrecoverable error.", exception: e.class, message: e.message, path: path, backtrace: e.backtrace)
+      rescue => e
+        # When the retry limit is reached or another error happen we will wait and retry.
+        #
+        # Thread might be stuck here, but I think its better than losing anything
+        # its either a transient errors or something bad really happened.
+        @logger.error('Deleting failed, retrying.', exception: e.class, message: e.message, path: path, backtrace: e.backtrace)
+        sleep RETRY_DELAY_SECONDS
```

Comment:
Add maximum number of retries, as we're not dealing with data loss (but duplication) here, and it's not worth an infinite loop. Might even want a limit on number of ingestion retries.

Reply:
Added the max retry number. Since I'm new to Ruby, if the changes not following the good practice, can you do the fix in your later changes?

Reply:
I've never written a line of Ruby code in my life :-)

Reply:
This looks like it has a good example to sue the Ruby "retry" feature: https://blog.appsignal.com/2018/05/16/ensure-retry-and-reraise-exceptions-in-ruby.html

Reply:
Thanks. Updated the code.